First Time User Experience

MelissaAutumn opened the pull request:
Wow it's done! I just had to battle with tailwind a bit.

This is a huge PR as it includes new components, fonts, assets based off the newly established design language for pro services. As well as slow switch away from tailwind. 

I've looked at this too long, so I'm sure there's some obvious inconsistencies I've missed. Please point them out!

Edit: Added some mobile screenshots. The ui is passable on mobile, but will need some design love in the future.

So here's some notes:

### TBPro
* Added BaseButton and its states to be used as a `button` would.
* Added PrimaryButton and SecondaryButton aliases
* Added TextInput and its states to be used as a `input` would.
* Added SelectInput and its states to be used as a `select` would.
* Added SyncCard which is a fancy checkbox list.
* Added Tooltip which is a WIP because it turns out I'm bad at coding tooltips.
* Added NoticeBar for general information with various states.
* Added Metropolis and Inter fonts.
* Added css variables to colors and sizing.
* Added BubbleSelect component.

### Additions
* Added ftue_level to Subscribers. So we can potentially run them through a FTUE wizard after major changes in the future. For now when the FTUE is done its set to 1.
* Added post-css's custom media plugin to variable-ize media query breakpoints.
* Force FTUE to be in light mode (as dark mode isn't ready yet.)
* Added FTUE with a store handling routing.

### Fixes
* The "My Link" is now properly url-encoded.
* Added some field validation to Schedule and Subscriber schemas.



Screenshots:
![Screen Shot 2024-06-26 at 12 47 02-fullpage](https://github.com/thunderbird/appointment/assets/97147377/596256dc-8df6-4cac-936d-ac6121b8610b)
![Screen Shot 2024-06-26 at 12 47 07-fullpage](https://github.com/thunderbird/appointment/assets/97147377/fd4f6e77-e835-46d2-9ad8-de2e6cf56692)
![Screen Shot 2024-06-26 at 12 48 13-fullpage](https://github.com/thunderbird/appointment/assets/97147377/15631889-b657-4c0b-9c5f-e810e4e78ac3)
![Screen Shot 2024-06-26 at 12 48 45-fullpage](https://github.com/thunderbird/appointment/assets/97147377/37a73c3f-f7ef-4b8e-909d-9dd5a452c19c)
![Screen Shot 2024-06-26 at 12 48 56-fullpage](https://github.com/thunderbird/appointment/assets/97147377/5cc772ce-e690-4534-8210-2762d22c3549)
![Screen Shot 2024-06-26 at 12 49 02-fullpage](https://github.com/thunderbird/appointment/assets/97147377/6d419c26-709a-4e78-9b81-f3e53ecb2f68)

Mobile screenshots:

![Screen Shot 2024-06-26 at 13 31 22-fullpage](https://github.com/thunderbird/appointment/assets/97147377/d4597f7b-ce5d-482b-a3dc-d27efa29f3f0)
![Screen Shot 2024-06-26 at 13 31 26-fullpage](https://github.com/thunderbird/appointment/assets/97147377/b15ac17e-e042-4f62-8230-aa3c6d9380fa)
![Screen Shot 2024-06-26 at 13 32 11-fullpage](https://github.com/thunderbird/appointment/assets/97147377/3ba4e6cd-4bc2-4066-a7d4-95182c1419e1)
![Screen Shot 2024-06-26 at 13 32 17-fullpage](https://github.com/thunderbird/appointment/assets/97147377/47e16527-1b80-4b58-a766-0279714f549f)
![Screen Shot 2024-06-26 at 13 32 21-fullpage](https://github.com/thunderbird/appointment/assets/97147377/46a6723e-ab73-4bef-ac0b-32188dc8d501)
![Screen Shot 2024-06-26 at 13 32 25-fullpage](https://github.com/thunderbird/appointment/assets/97147377/c7e446fd-9647-4476-b08f-e910bfda840e)

devmount commented on backend/src/appointment/migrations/versions/2024_06_25_2133-156b3b0d77b9_add_ftue_level_to_subscribers.py:
```diff
@@ -0,0 +1,24 @@
+
+
+# revision identifiers, used by Alembic.
+revision = '156b3b0d77b9'
+down_revision = '12c7e1b34dd6'
```
Unfortunately this collides with previous developments 😬 
We had two migrations on 2024-06-13 which change the current migration head. I'll fix the head locally on my system, but we should provide a continuous migration queue.

MelissaAutumn replied:
Yeah I haven't had the chance to rebase this off of main yet. I'll create a merge migration.

I'm not a fan of git-based database migrations for this reason :sweat_smile:

devmount replied:
Yeah, this is a pain during development... Thanks for a merge migration 🙏🏻

devmount reviewed (commented):
First of all: Thank you for this great work! It does not only implement FTUE, but also provides the base for the upcoming changes like the transition away from Tailwind 👏🏻 I feel like we should start documentation on our coding conventions (like which order we use for .vue file sections - script/template/css, how CSS is structure etc.).

But now for the review: I found a lot of small stuff, let me know if I should fix them right away. There is one issue with db migrations, and I currently have issues running the frontend container. I get the following error:

![image](https://github.com/thunderbird/appointment/assets/5441654/614e281c-5015-43a8-8ccf-50c04122feca)

Do I need to rebuild the whole container from scratch?

MelissaAutumn commented on backend/src/appointment/routes/api.py:
```diff
@@ -1,3 +1,4 @@
+import logging
```
There's a log statement in api.py that was missing this import

devmount replied:
Oh my bad! Thanks!

MelissaAutumn commented on frontend/tailwind.config.js:
```diff
@@ -1,4 +1,5 @@
 const defaultTheme = require('tailwindcss/defaultTheme');
+const plugin = require('tailwindcss/plugin.js');
```
Nothing anymore :scream:

MelissaAutumn commented:
For the build stuff you'll have to blow away your node_module volume and rebuild the container. I'm not sure how other projects handle dockerized node_modules. :thinking:

devmount commented on frontend/src/locales/en.json:
```diff
@@ -120,6 +120,7 @@
     "busy": "Busy",
     "caldav": "CalDAV",
     "calendar": "Calendar",
+    "calendars": "Calendars",
```
Should we rather use i18n plural functionality here?

MelissaAutumn replied:
Good catch!

devmount reviewed (approved):
FTUE works on my end now! I love the design! Can't wait to implement it for the whole application 🙏🏻 

Only thing I found: After logging in, I have a bug causing input fields to be dark, but everything else light. This was fixed be reloading the page, so already having some local storage set might be the reason here.
    ![image](https://github.com/thunderbird/appointment/assets/5441654/d43a8db4-8154-4290-9590-28dac51e5ff8)

Also I added German translations.

devmount commented on frontend/src/tbpro/elements/BaseButton.vue:
```diff
@@ -0,0 +1,147 @@
+<template>
+  <button :class="{'primary': type === 'primary', 'secondary': type === 'secondary', 'small': size === 'small'}" type="button">
```
Alright, makes sense!

MelissaAutumn commented:
@devmount I added a new string!

devmount commented on frontend/src/stores/external-connections-store.ts:
```diff
@@ -1,6 +1,7 @@
-import { ExternalConnection, ExternalConnectionCollection, Fetch, ExternalConnectionCollectionResponse } from '@/models';
-import { defineStore } from 'pinia';
-import { ref, computed } from 'vue';
+import {ExternalConnection, ExternalConnectionCollection, Fetch, ExternalConnectionCollectionResponse} from '@/models';
+import {defineStore} from 'pinia';
+import {ref, computed} from 'vue';
```
Is this a linter thing to remove spaces around curly braces?

devmount commented on frontend/src/stores/schedule-store.ts:
```diff
@@ -33,7 +37,7 @@ export const useScheduleStore = defineStore('schedules', () => {
 
     const { updateScheduleUrls } = useUserStore();
-    updateScheduleUrls(data.value);
+    await updateScheduleUrls(data.value);
```
> 'await' has no effect on the type of this expression. - ts(80007)

You cannot await a method that is not async.

devmount commented on frontend/src/stores/schedule-store.ts:
```diff
@@ -1,10 +1,14 @@
 
 // eslint-disable-next-line import/prefer-default-export
 export const useScheduleStore = defineStore('schedules', () => {
+  const dj = inject('dayjs');
```
This needs to be typed, otherwise TS won't know that dj is a Dayjs callable

```suggestion
  const dj = inject<(date?: ConfigType) => Dayjs>('dayjs');
```

with

```js
import { Dayjs, ConfigType } from 'dayjs';
```

devmount reviewed (commented):
Thanks for merging the TS changes in 👏🏻  Left some comments about possible issues I found.

MelissaAutumn commented:
Couple of things:
* Either I broke the eslint config or it wasn't setup for typescript, so I fixed that
* That then broke my ide's "Run eslint on save" functionality, oops
* I fixed that, and then got annoyed at how we type injections
* Switched dayjs to a typed Symbol
* Got annoyed it wasn't providing timezone
* Added a new type that's an instance of Dayjs class with tz mixed in maybe?
* idk it works now, we should maybe switch off dayjs in the future

Works for me, but I'll merge this onto stage and we can give it a good test.
